sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,1525 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Dorian Fevrier, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+#define APPLESEED_ENABLE_IMATH_INTEROP
+
+// appleseed.renderer headers.
+#include "renderer/api/camera.h"
+#include "renderer/api/log.h"
+#include "renderer/api/object.h"
+#include "renderer/api/project.h"
+#include "renderer/api/scene.h"
+#include "renderer/api/utility.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/matrix.h"
+#include "foundation/math/triangulator.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/autoreleaseptr.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/string.h"
+
+// appleseed.main headers.
+#include "main/dllvisibility.h"
+
+// Alembic headers.
+#include <Alembic/AbcCoreFactory/All.h>
+#include <Alembic/AbcGeom/All.h>
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+#include <vector>
+#include <algorithm>
+
+namespace
+{
+    // shortcuts to save space
+    namespace asf = foundation;
+    namespace asr = renderer;
+
+    namespace Abc        = Alembic::Abc;
+    namespace AbcCoreFac = Alembic::AbcCoreFactory;
+    namespace AbcGeom    = Alembic::AbcGeom;
+    namespace AbcUtil    = Alembic::Util;
+
+    typedef Alembic::Abc::index_t  AbcIndex_t;
+    typedef Alembic::Abc::chrono_t AbcChrono_t;
+
+    // roll into given matrix stack and return the flatten one
+    asf::Matrix4d flatten_xform(const std::vector<asf::Matrix4d>& mtx_stack)
+    {
+        // the matrix we will return
+        asf::Matrix4d out_mtx = asf::Matrix4d::identity();
```

Comment:
A small optimization would be to initialize `out_mtx` with the first matrix from the stack (but then you can't use a range-based `for` loop).

Reply:
Done.